cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamEventBuffer.java

```diff
@@ -0,0 +1,134 @@
+
+  StreamEventBuffer() {
+    this.buffer = ByteBuffers.EMPTY_BUFFER;
+    this.bufferInput = new ByteBufferInputStream(buffer);
+    this.decoder = new BinaryDecoder(bufferInput);
+    this.basePosition = -1L;
+  }
+
+  /**
+   * Fills the internal buffer by reading from the given input stream.
+   *
+   * @param input input stream to read from
+   * @param size number of bytes to read
+   * @throws IOException if failed to read from the stream
+   * @throws EOFException if failed to read the given number of bytes from the input
+   */
+  void fillBuffer(SeekableInputStream input, int size) throws IOException {
+    buffer.clear();
+    buffer = ensureCapacity(buffer, size);
+
+    try {
+      basePosition = input.getPos();
+      int bytesRead = input.read(buffer.array(), 0, size);
+      if (bytesRead != size) {
+        throw new EOFException("Expected to read " + size + ", but only " + bytesRead + " was read");
+      }
+      buffer.limit(size);
+      bufferInput.reset(buffer);
+    } catch (IOException e) {
+      // Make the buffer has nothing to read
+      buffer.position(buffer.limit());
+      basePosition = -1L;
+      throw e;
+    }
+  }
+
+  /**
+   * Returns {@code true} if there are events in the buffer, {@code false} otherwise.
+   */
+  boolean hasEvent() {
+    return buffer.hasRemaining();
+  }
+
+  /**
+   * Returns the position in the stream that this buffer is currently at or {@code -1} if nothing has been
+   * read from the stream.
+   */
+  long getPosition() {
+    return basePosition >= 0 ? basePosition + buffer.position() : -1L;
+  }
+
+  /**
+   * Returns the position in the stream that represents the end of this buffer or {@code -1} if nothing has
+   * been read from the stream.
+   */
+  long getEndPosition() {
+    return basePosition >= 0 ? basePosition + buffer.limit() : -1L;
+  }
+
+  /**
+   * Decodes a stream event from the buffer.
+   *
+   * @param timestamp timestamp of the {@link PositionStreamEvent} created
+   * @param filter filter to apply to decide reading or skipping event
+   * @return A {@link PositionStreamEvent} if the filter accept it, or {@code null} if rejected by the filter
+   * @throws IOException if fails to decode event from the buffer
+   */
+  PositionStreamEvent nextEvent(long timestamp, ReadFilter filter) throws IOException {
+    if (!hasEvent()) {
+      throw new IOException("No more event in the buffer");
+    }
+
+    long eventPos = basePosition + buffer.position();
+    if (filter.acceptOffset(eventPos)) {
+      return new PositionStreamEvent(StreamEventDataCodec.decode(decoder), timestamp, eventPos);
+    }
+    StreamEventDataCodec.skip(decoder);
+    return null;
+  }
+
+  /**
+   * Ensures the given {@link ByteBuffer} has enough size.
+   *
+   * @param buffer The buffer to check for capacity
+   * @param size Capacity needed
+   * @return The given buffer if it has enough space. Otherwise a new buffer of the given size will be returned.
```

Comment:
The given buffer if it is of sufficient size; otherwise, a new buffer of the given size will be returned.